### PR TITLE
Adds initial dtls 1.3 structs and definitions

### PR DESCRIPTION
**Pull request by fwh-dc**
My idea is that this should form the base of the dtls 1.3 feature branch as described in https://github.com/openssl/openssl/issues/13900

### CODE REVIEW

File: ssl/methods.c
Comment (mattcaswell) [inline ssl/methods.c:128]: This guard makes no sense because we will not have a public DTLSv1_3_method() function. We don't have this for the TLSv1.3 equivalent.

File: ssl/methods.c
Comment (mattcaswell) [inline ssl/methods.c:154]: and here

File: ssl/methods.c
Comment (mattcaswell) [inline ssl/methods.c:184]: and here

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: include/openssl/ssl.h.in
Comment (fwh-dc) [reply include/openssl/ssl.h.in:408]: Hmm i was trying to follow what seemed to be a convention:

```SSL_OP_NO_TLSv1``` == ```SSL_OP_NO_DTLSv1```
```SSL_OP_NO_TLSv1_2``` == ```SSL_OP_NO_DTLSv1_2```

Comment (mattcaswell) [reply include/openssl/ssl.h.in:408]: Ah, yes. You are right. So, ignore this then.

File: ssl/methods.c
Comment (fwh-dc) [reply ssl/methods.c:128]: Cool will fix it.

**Comment by fwh-dc** [comment]
@mattcaswell 
Just wanted to make sure that you're not waiting for an update from my side. Can you give an update?

**Review by mattcaswell — APPROVED** [review]

**Comment by mattcaswell** [comment]
Sorry, things are currently very busy due to work on the forthcoming 3.2 beta. My time to look at this is limited at the moment.

I've placed an OMC hold on this because we still don't have a branch set up for this yet. That is likely to take a while I think. That shouldn't stop reviews of this PR going ahead.

**Comment by fwh-dc** [comment]
Understandable that other tasks takes priority. Would you prefer that I push the rest of the changes as PRs against master or should I wait until we have the feature branch setup?

**Comment by mattcaswell** [comment]
> Understandable that other tasks takes priority. Do you prefer that I push the rest of the changes as PRs against master or should I wait until we have the feature branch setup?

I suggest creating a "DTLSv1.3" PR which collects together all the commits from your other PRs. You'll need to rebase it against master periodically. We wouldn't do any review in the DTLSv1.3 PR - we would do review in the the other PRs you create. But you can use that PR as the basis for new PRs. That does mean the other PRs will also have all the commits from the "DTLSv1.3" PR - which will be slightly confusing - but we just need to be very clear about which commits are being reviewed and which are not.

I hope I explained that well enough.

**Comment by fwh-dc** [comment]
Ok I think I get the idea. I think I will create PR's on the remaining "early development" fixes and target master, and then from those I will create the DTLS 1.3 pr branch which can be used to co-develop from if there's anyone in the community is interested in participating.

**Comment by mattcaswell** [comment]
Ping @openssl/committers for second review. The hold is only due to the lack of a branch to put this...it should not hold back reviews of this PR.

**Comment by mattcaswell** [comment]
Ping @openssl/committers

**Comment by mattcaswell** [comment]
Still looking for a second reviewer on this. The changes are quite straightforward. The hold is only there because we are missing a branch to push this to - but it shouldn't hold up review.

Ping @openssl/committers.

**Review by nhorman — APPROVED** [review]
Unsure what the check_docs failure was but it doesn't seem to recur when built locally, so I assume they've been fixed, or that it was transient

**Comment by mattcaswell** [comment]
Lifting the OMC hold on this, because we now have a feature branch

**Comment by mattcaswell** [comment]
Darn. So near and yet so far - repo is frozen for the alpha:

````
$ git push
Enumerating objects: 31, done.
Counting objects: 100% (31/31), done.
Delta compression using up to 20 threads
Compressing objects: 100% (18/18), done.
Writing objects: 100% (18/18), 1.88 KiB | 1.88 MiB/s, done.
Total 18 (delta 17), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (17/17), completed with 13 local objects.
remote: frozen.sh: failed with exit status 1
remote: [frozen] this repository is frozen in favor of: levitte
To github.openssl.org:openssl/openssl.git
 ! [remote rejected]       feature/dtls-1.3 -> feature/dtls-1.3 (pre-receive hook declined)
````

**Comment by mattcaswell** [comment]
@levitte would you be able to merge this to the feature/dtls-1.3 branch?

**Comment by levitte** [comment]
> @levitte would you be able to merge this to the feature/dtls-1.3 branch?

Yup, will do after lunch

**Comment by levitte** [comment]
Merged onto feature/dtls-1.3

84ce747b1cddc1b92eabf258431664b71d73df28 Adds initial dtls 1.3 structs and definitions
ca4f79aa4fd801aeb1116a9e36755edf0243f50e Remove compile guards for dtls1.3 method implementations

**Pull request closed without merging**
